packetbeat/protos/sip/parser.go

```diff
@@ -0,0 +1,514 @@
+		headerName := getExpandedHeaderName(toLower(headerNameBuf[:], data[:i]))
+		headerVal := trim(data[i+1 : p])
+		if isDebug {
+			debugf("Header: '%s' Value: '%s'\n", data[:i], headerVal)
+		}
+
+		// Headers we need for parsing. Make sure we always
+		// capture their value
+		switch {
+		case bytes.Equal(headerName, nameMaxForwards):
+			m.maxForwards, _ = parseInt(headerVal)
+		case bytes.Equal(headerName, nameContentLength):
+			m.contentLength, _ = parseInt(headerVal)
+			m.hasContentLength = true
+		case bytes.Equal(headerName, nameContentType):
+			m.contentType = headerVal
+		case bytes.Equal(headerName, nameUserAgent):
+			m.userAgent = headerVal
+		case bytes.Equal(headerName, nameTo):
+			m.to = headerVal
+		case bytes.Equal(headerName, nameFrom):
+			m.from = headerVal
+		case bytes.Equal(headerName, nameCseq):
+			m.cseq = headerVal
+		case bytes.Equal(headerName, nameCallID):
+			m.callID = headerVal
+		case bytes.Equal(headerName, nameAllow):
+			m.allow = parseCommaSeparatedList(headerVal)
+		case bytes.Equal(headerName, nameSupported):
+			m.supported = parseCommaSeparatedList(headerVal)
+		case bytes.Equal(headerName, nameVia):
+			m.via = append(m.via, headerVal)
+		}
+
+		m.headers[string(headerName)] = append(
+			m.headers[string(headerName)],
+			headerVal,
+		)
+
+		return p + 2, nil
+	}
+
+	return len(data), nil
+}
+
+func parseCommaSeparatedList(s common.NetString) (list []string) {
+	values := bytes.Split(s, []byte(","))
+	list = make([]string, len(values))
+	for idx := range values {
+		list[idx] = string(bytes.ToLower(bytes.Trim(values[idx], " ")))
+	}
+	return list
+}
+
+func (*parser) parseBody(pi *parsingInfo, m *message) {
+	nbytes := len(pi.data)
+	if nbytes >= m.contentLength-pi.bodyReceived {
+		wanted := m.contentLength - pi.bodyReceived
+		m.body = append(m.body, pi.data[:wanted]...)
+		pi.bodyReceived = m.contentLength
+		m.size += uint64(wanted)
+		pi.data = pi.data[wanted:]
+	} else {
+		m.body = append(m.body, pi.data...)
+		pi.data = nil
+		pi.bodyReceived += nbytes
+		m.size += uint64(nbytes)
+		if isDebug {
+			debugf("bodyReceived: %d", pi.bodyReceived)
+		}
+	}
+}
+
+func (m *message) getEndpoints() (src *common.Endpoint, dst *common.Endpoint) {
+	source, destination := common.MakeEndpointPair(m.ipPortTuple.BaseTuple, m.cmdlineTuple)
+	src, dst = &source, &destination
+	return src, dst
+}
+
+func trim(buf []byte) []byte {
+	return trimLeft(trimRight(buf))
+}
+
+func trimLeft(buf []byte) []byte {
+	for i, b := range buf {
+		if b != ' ' && b != '\t' {
+			return buf[i:]
+		}
+	}
+	return nil
+}
+
+func trimRight(buf []byte) []byte {
+	for i := len(buf) - 1; i > 0; i-- {
+		b := buf[i]
+		if b != ' ' && b != '\t' {
+			return buf[:i+1]
+		}
+	}
+	return nil
+}
+
+func parseInt(line []byte) (int, error) {
+	buf := streambuf.NewFixed(line)
+	i, err := buf.IntASCII(false)
+	return int(i), err
+	// TODO: is it an error if 'buf.Len() != 0 {}' ?
+}
+
+func toLower(buf, in []byte) []byte {
```

Comment:
Is this really needed? It looks like this is trying to potentially avoid a slice allocation by modifying what might be (I haven't traced through the code fully) a reusable byte slice, `buf`, in-place. Are we worried that the volume of allocations might be so high in this parser that we'll cause extra GC/Memory overhead that impacts overall packetbeat performance? I would imagine that we wouldn't require this parser to be _so_ performance intensive where this would actually matter much...